lib/Synchronise/Store.hs

```diff
@@ -0,0 +1,304 @@
+{-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE RankNTypes      #-}
+
+module Synchronise.Store where
+
+import Control.Lens
+import Data.ByteString (ByteString)
+
+import Synchronise.Diff
+import Synchronise.Document
+import Synchronise.Identifier
+
+type DiffID  = Int
+type OpID    = Int
+
+data ConflictResp = ConflictResp
+  { _conflictRawDoc  :: ByteString
+  , _conflictRawDiff :: ByteString
+  , _conflictDiffID  :: DiffID
+  , _conflictRawOps  :: [(OpID, ByteString)]
+  }
+makeLenses ''ConflictResp
+
+data DiffResp = DiffResp
+  { _diffEntity    :: ByteString
+  , _diffKey       :: Int
+  , _diffPatch     :: LabelledPatch ()
+  , _diffConflicts :: [LabelledPatch ()]
+  }
+makeLenses ''DiffResp
+
+data OpResp label = OpResp 
+  { _opDiffID :: DiffID
+  , _opID     :: OpID
+  , _ops      :: LabelledOp label
+  }
+makeLenses ''OpResp
+
+-- | The internal store "module".
+--
+data Store store = Store
+  { -- | Initialise a handle to the storage backend.
+    initBackend  :: store -> IO ()
+
+    -- | Finalise a handle to the storage backend.
+  , closeBackend :: store -> IO ()
+
+    -- | Brackets an action with init/finalise store.
+  , withStore    :: store -> IO a -> IO a
+
+    -- | clone a store handle for concurrent operations. the new handle must be
+    -- safe to use concurrently with the original handle.
+    --
+    -- (e.g. open a second connection to the same postgresql database.)
+    --
+    -- this may be a no-op for backends which are already thread-safe (e.g.
+    -- in-memory iorefs).
+  , cloneStore   :: store -> IO (StoreOp store)
+
+    -- Operations on interal keys
+
+    -- | Allocate and return a new 'InternalKey'.
+  , createInternalKey :: store -> IO InternalKey
+
+    -- | Find the 'InternalKey' associated with a 'ForeignKey'.
+  , lookupInternalKey :: store -> ForeignKey  -> IO (Maybe InternalKey)
+
+    -- | Delete an 'InternalKey' and any associated resources.
+  , deleteInternalKey :: store -> InternalKey -> IO Int
+
+
+    -- Operations on foreign keys
+
+    -- | Record a 'ForeignKey' and it's association with an 'InternalKey'.
+  , recordForeignKey :: store -> InternalKey -> ForeignKey -> IO ()
+
+    -- | Find the 'ForeignKey' corresponding to an 'InternalKey'
+  , lookupForeignKey :: store -> InternalKey -> IO (Maybe ForeignKey)
```

Comment:
This won't work like this any more; it'll need to take a "data source" argument. In retcon it figured out which data source to look for with ScopedTypeVariables from the phantom type param in the return type.